From 7130412929e7bd5595f67d92c91cb1b1ee6475c0 Mon Sep 17 00:00:00 2001
From: Richard W.M. Jones <rjones@redhat.com>
Date: Wed, 26 Jan 2011 23:39:40 +0000
Subject: [PATCH 19/20] Confirm Microsoft Hyper-V, and add a regression test.

Data collected from a RHEL 5.6 guest running on Hyper-V.

Also this modifies the VirtualPC test slightly to distinguish
it from Hyper-V.  We require more data from a real VirtualPC
guest to properly refine this test.
(cherry picked from commit 0e90af4014da8065a4062319921288baf972c6fb)
---
 Makefile.am                              |    7 +
 tests/hyperv/proc/cpuinfo                |   24 +
 tests/hyperv/proc/self/status            |   36 ++
 tests/hyperv/sbin/dmidecode              |  965 ++++++++++++++++++++++++++++++
 tests/hyperv/sbin/id                     |    2 +
 tests/hyperv/sbin/virt-what-cpuid-helper |    3 +
 tests/test-hyperv.sh                     |   34 +
 virt-what.in                             |    6 +-
 virt-what.pod                            |    2 +-
 9 files changed, 1076 insertions(+), 3 deletions(-)
 create mode 100644 tests/hyperv/proc/cpuinfo
 create mode 100644 tests/hyperv/proc/self/status
 create mode 100755 tests/hyperv/sbin/dmidecode
 create mode 100755 tests/hyperv/sbin/id
 create mode 100755 tests/hyperv/sbin/virt-what-cpuid-helper
 create mode 100755 tests/test-hyperv.sh

diff --git a/Makefile.am b/Makefile.am
index 9601be9..0e65023 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -28,6 +28,11 @@ EXTRA_DIST = \
 	tests/esx4.1/sbin/dmidecode \
 	tests/esx4.1/sbin/id \
 	tests/esx4.1/sbin/virt-what-cpuid-helper \
+	tests/hyperv/proc/cpuinfo \
+	tests/hyperv/proc/self/status \
+	tests/hyperv/sbin/dmidecode \
+	tests/hyperv/sbin/id \
+	tests/hyperv/sbin/virt-what-cpuid-helper \
 	tests/kvm/proc/cpuinfo \
 	tests/kvm/proc/self/status \
 	tests/kvm/sbin/dmidecode \
@@ -98,6 +103,7 @@ EXTRA_DIST = \
 	tests/zvm/sbin/virt-what-cpuid-helper \
 	tests/test-baremetal.sh \
 	tests/test-esx4.1.sh \
+	tests/test-hyperv.sh \
 	tests/test-kvm.sh \
 	tests/test-qemu.sh \
 	tests/test-rhel5-xen-dom0.sh \
@@ -124,6 +130,7 @@ virt-what.txt: virt-what.pod
 TESTS = \
 	tests/test-baremetal.sh \
 	tests/test-esx4.1.sh \
+	tests/test-hyperv.sh \
 	tests/test-kvm.sh \
 	tests/test-qemu.sh \
 	tests/test-rhel5-xen-dom0.sh \
diff --git a/tests/hyperv/proc/cpuinfo b/tests/hyperv/proc/cpuinfo
new file mode 100644
index 0000000..0559ab4
--- /dev/null
+++ b/tests/hyperv/proc/cpuinfo
@@ -0,0 +1,24 @@
+processor	: 0
+vendor_id	: GenuineIntel
+cpu family	: 6
+model		: 15
+model name	: Intel(R) Xeon(R) CPU            5130  @ 2.00GHz
+stepping	: 6
+cpu MHz		: 1977.298
+cache size	: 4096 KB
+physical id	: 0
+siblings	: 1
+core id		: 0
+cpu cores	: 1
+apicid		: 0
+fpu		: yes
+fpu_exception	: yes
+cpuid level	: 6
+wp		: yes
+flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx lm constant_tsc up pni ssse3 cx16 lahf_lm
+bogomips	: 3954.59
+clflush size	: 64
+cache_alignment	: 64
+address sizes	: 40 bits physical, 48 bits virtual
+power management:
+
diff --git a/tests/hyperv/proc/self/status b/tests/hyperv/proc/self/status
new file mode 100644
index 0000000..fd4a88e
--- /dev/null
+++ b/tests/hyperv/proc/self/status
@@ -0,0 +1,36 @@
+Name:	cat
+State:	R (running)
+SleepAVG:	88%
+Tgid:	4747
+Pid:	4747
+PPid:	3409
+TracerPid:	0
+Uid:	0	0	0	0
+Gid:	0	0	0	0
+FDSize:	256
+Groups:	0 1 2 3 4 6 10 
+VmPeak:	   58944 kB
+VmSize:	   58944 kB
+VmLck:	       0 kB
+VmHWM:	     528 kB
+VmRSS:	     528 kB
+VmData:	     180 kB
+VmStk:	      88 kB
+VmExe:	      20 kB
+VmLib:	    1448 kB
+VmPTE:	      36 kB
+StaBrk:	19d39000 kB
+Brk:	19d5a000 kB
+StaStk:	7fff02c16930 kB
+Threads:	1
+SigQ:	1/8191
+SigPnd:	0000000000000000
+ShdPnd:	0000000000000000
+SigBlk:	0000000000000000
+SigIgn:	0000000000000000
+SigCgt:	0000000000000000
+CapInh:	0000000000000000
+CapPrm:	00000000fffffeff
+CapEff:	00000000fffffeff
+Cpus_allowed:	00000000,00000000,00000000,00000000,00000000,00000000,7fffffff,ffffffff
+Mems_allowed:	00000000,00000001
diff --git a/tests/hyperv/sbin/dmidecode b/tests/hyperv/sbin/dmidecode
new file mode 100755
index 0000000..d126757
--- /dev/null
+++ b/tests/hyperv/sbin/dmidecode
@@ -0,0 +1,965 @@
+#!/bin/sh -
+cat <<'EOF'
+# dmidecode 2.10
+SMBIOS 2.3 present.
+67 structures occupying 3752 bytes.
+Table at 0x000F8850.
+
+Handle 0x0000, DMI type 0, 20 bytes
+BIOS Information
+	Vendor: American Megatrends Inc.
+	Version: 080002 
+	Release Date: 05/05/2008
+	Address: 0xF0000
+	Runtime Size: 64 kB
+	ROM Size: 256 kB
+	Characteristics:
+		ISA is supported
+		PCI is supported
+		PNP is supported
+		BIOS is upgradeable
+		BIOS shadowing is allowed
+		ESCD support is available
+		Boot from CD is supported
+		Selectable boot is supported
+		BIOS ROM is socketed
+		EDD is supported
+		5.25"/360 kB floppy services are supported (int 13h)
+		5.25"/1.2 MB floppy services are supported (int 13h)
+		3.5"/720 kB floppy services are supported (int 13h)
+		3.5"/2.88 MB floppy services are supported (int 13h)
+		Print screen service is supported (int 5h)
+		8042 keyboard services are supported (int 9h)
+		Serial services are supported (int 14h)
+		Printer services are supported (int 17h)
+		CGA/mono video services are supported (int 10h)
+		AGP is supported
+		LS-120 boot is supported
+		ATAPI Zip drive boot is supported
+		BIOS boot specification is supported
+
+Handle 0x0001, DMI type 1, 25 bytes
+System Information
+	Manufacturer: Microsoft Corporation
+	Product Name: Virtual Machine
+	Version: 5.0
+	Serial Number: 9531-7949-1919-0512-8596-0146-27
+	UUID: 87123C54-42BA-694D-90FF-FD3B05DCDB42
+	Wake-up Type: Power Switch
+
+Handle 0x0002, DMI type 2, 8 bytes
+Base Board Information
+	Manufacturer: Microsoft Corporation
+	Product Name: Virtual Machine
+	Version: 5.0
+	Serial Number: 9531-7949-1919-0512-8596-0146-27
+
+Handle 0x0003, DMI type 3, 17 bytes
+Chassis Information
+	Manufacturer: Microsoft Corporation
+	Type: Desktop
+	Lock: Not Present
+	Version: 5.0
+	Serial Number: 9531-7949-1919-0512-8596-0146-27
+	Asset Tag: 4703-2669-3135-1081-7372-7632-77
+	Boot-up State: Safe
+	Power Supply State: Safe
+	Thermal State: Other
+	Security Status: Other
+	OEM Information: 0x00000000
+
+Handle 0x0004, DMI type 11, 5 bytes
+OEM Strings
+	String 1: [MS_VM_CERT/SHA1/9b80ca0d5dd061ec9da4e494f4c3fd1196270c22]
+	String 2: 00000000000000000000000000000000
+	String 3: To be filed by MSFT
+
+Handle 0x0005, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_1
+	Type: Central Processor
+	Family: Pentium III
+	Manufacturer: Intel            
+	ID: F6 06 00 00 FF FB 8B 1F
+	Signature: Type 0, Family 6, Model 15, Stepping 6
+	Flags:
+		FPU (Floating-point unit on-chip)
+		VME (Virtual mode extension)
+		DE (Debugging extension)
+		PSE (Page size extension)
+		TSC (Time stamp counter)
+		MSR (Model specific registers)
+		PAE (Physical address extension)
+		MCE (Machine check exception)
+		CX8 (CMPXCHG8 instruction supported)
+		APIC (On-chip APIC hardware supported)
+		SEP (Fast system call)
+		MTRR (Memory type range registers)
+		PGE (Page global enable)
+		MCA (Machine check architecture)
+		CMOV (Conditional move instruction supported)
+		PAT (Page attribute table)
+		PSE-36 (36-bit page size extension)
+		CLFSH (CLFLUSH instruction supported)
+		MMX (MMX technology supported)
+		FXSR (Fast floating-point save and restore)
+		SSE (Streaming SIMD extensions)
+		SSE2 (Streaming SIMD extensions 2)
+		SS (Self-snoop)
+		HTT (Hyper-threading technology)
+	Version: Genuine Intel(R) Processor              
+	Voltage: 3.3 V 2.9 V
+	External Clock: 66 MHz
+	Max Speed: 333 MHz
+	Current Speed: 333 MHz
+	Status: Populated, Enabled
+	Upgrade: 370-pin Socket
+	L1 Cache Handle: 0x0006
+	L2 Cache Handle: 0x0007
+	L3 Cache Handle: 0x0008
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_1
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_1
+	Part Number: MKF_PROCESSOR_PART_NO_1
+
+Handle 0x0006, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0007, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0008, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0009, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_2
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x000A
+	L2 Cache Handle: 0x000B
+	L3 Cache Handle: 0x000C
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_2
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_2
+	Part Number: MKF_PROCESSOR_PART_NO_2
+
+Handle 0x000A, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x000B, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x000C, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x000D, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_3
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x000E
+	L2 Cache Handle: 0x000F
+	L3 Cache Handle: 0x0010
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_3
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_3
+	Part Number: MKF_PROCESSOR_PART_NO_3
+
+Handle 0x000E, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x000F, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0010, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0011, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_4
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x0012
+	L2 Cache Handle: 0x0013
+	L3 Cache Handle: 0x0014
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_4
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_4
+	Part Number: MKF_PROCESSOR_PART_NO_4
+
+Handle 0x0012, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0013, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0014, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0015, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_5
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x0016
+	L2 Cache Handle: 0x0017
+	L3 Cache Handle: 0x0018
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_5
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_5
+	Part Number: MKF_PROCESSOR_PART_NO_5
+
+Handle 0x0016, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0017, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0018, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0019, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_6
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x001A
+	L2 Cache Handle: 0x001B
+	L3 Cache Handle: 0x001C
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_6
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_6
+	Part Number: MKF_PROCESSOR_PART_NO_6
+
+Handle 0x001A, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x001B, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x001C, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x001D, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_7
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x001E
+	L2 Cache Handle: 0x001F
+	L3 Cache Handle: 0x0020
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_7
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_7
+	Part Number: MKF_PROCESSOR_PART_NO_7
+
+Handle 0x001E, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x001F, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0020, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0021, DMI type 4, 35 bytes
+Processor Information
+	Socket Designation: MKF_PROCESSOR_SOCKET_DESIGNATION_8
+	Type: Other
+	Family: Other
+	Manufacturer:                  
+	ID: 00 00 00 00 00 00 00 00
+	Version:                                         
+	Voltage: 3.3 V 2.9 V
+	External Clock: Unknown
+	Max Speed: 500 MHz
+	Current Speed: Unknown
+	Status: Unpopulated
+	Upgrade: None
+	L1 Cache Handle: 0x0022
+	L2 Cache Handle: 0x0023
+	L3 Cache Handle: 0x0024
+	Serial Number: MKF_PROCESSOR_SERIAL_NO_8
+	Asset Tag: MKF_PROCESSOR_ASSET_TAG_8
+	Part Number: MKF_PROCESSOR_PART_NO_8
+
+Handle 0x0022, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L1-Cache
+	Configuration: Disabled, Not Socketed, Level 1
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0023, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L2-cache
+	Configuration: Disabled, Not Socketed, Level 2
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0024, DMI type 7, 19 bytes
+Cache Information
+	Socket Designation: L3-cache
+	Configuration: Disabled, Not Socketed, Level 3
+	Operational Mode: Unknown
+	Location: Internal
+	Installed Size: 0 kB
+	Maximum Size: 0 kB
+	Supported SRAM Types:
+		Unknown
+	Installed SRAM Type: Unknown
+	Speed: Unknown
+	Error Correction Type: Unknown
+	System Type: Unknown
+	Associativity: Unknown
+
+Handle 0x0025, DMI type 5, 24 bytes
+Memory Controller Information
+	Error Detecting Method: 64-bit ECC
+	Error Correcting Capabilities:
+		None
+	Supported Interleave: One-way Interleave
+	Current Interleave: One-way Interleave
+	Maximum Memory Module Size: 512 MB
+	Maximum Total Memory Size: 2048 MB
+	Supported Speeds:
+		70 ns
+		60 ns
+	Supported Memory Types:
+		Other
+		Unknown
+		Standard
+		FPM
+		EDO
+		Parity
+		ECC
+		SIMM
+		DIMM
+		Burst EDO
+		SDRAM
+	Memory Module Voltage: 3.3 V
+	Associated Memory Slots: 4
+		0x0026
+		0x0027
+		0x0028
+		0x0029
+	Enabled Error Correcting Capabilities:
+		None
+
+Handle 0x0026, DMI type 6, 12 bytes
+Memory Module Information
+	Socket Designation: DIMM1
+	Bank Connections: 0 0
+	Current Speed: Unknown
+	Type: EDO DIMM
+	Installed Size: 16 MB (Single-bank Connection)
+	Enabled Size: 16 MB (Single-bank Connection)
+	Error Status: OK
+
+Handle 0x0027, DMI type 6, 12 bytes
+Memory Module Information
+	Socket Designation: DIMM2
+	Bank Connections: 0 0
+	Current Speed: Unknown
+	Type: EDO DIMM
+	Installed Size: 16 MB (Single-bank Connection)
+	Enabled Size: 16 MB (Single-bank Connection)
+	Error Status: OK
+
+Handle 0x0028, DMI type 6, 12 bytes
+Memory Module Information
+	Socket Designation: DIMM3
+	Bank Connections: 0 0
+	Current Speed: Unknown
+	Type: EDO DIMM
+	Installed Size: 16 MB (Single-bank Connection)
+	Enabled Size: 16 MB (Single-bank Connection)
+	Error Status: OK
+
+Handle 0x0029, DMI type 6, 12 bytes
+Memory Module Information
+	Socket Designation: DIMM4
+	Bank Connections: 0 0
+	Current Speed: Unknown
+	Type: EDO DIMM
+	Installed Size: 16 MB (Single-bank Connection)
+	Enabled Size: 16 MB (Single-bank Connection)
+	Error Status: OK
+
+Handle 0x002A, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: USB
+	Internal Connector Type: Centronics
+	External Reference Designator: USB1
+	External Connector Type: Centronics
+	Port Type: USB
+
+Handle 0x002B, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: USB
+	Internal Connector Type: Centronics
+	External Reference Designator: USB2
+	External Connector Type: Centronics
+	Port Type: USB
+
+Handle 0x002C, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: COM1
+	Internal Connector Type: DB-9 female
+	External Reference Designator: COM1
+	External Connector Type: DB-9 female
+	Port Type: Serial Port 16550A Compatible
+
+Handle 0x002D, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: COM2
+	Internal Connector Type: DB-9 female
+	External Reference Designator: COM2
+	External Connector Type: DB-9 female
+	Port Type: Serial Port 16550A Compatible
+
+Handle 0x002E, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: Printer
+	Internal Connector Type: DB-25 male
+	External Reference Designator: Lpt1
+	External Connector Type: DB-25 male
+	Port Type: Parallel Port ECP/EPP
+
+Handle 0x002F, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: Video
+	Internal Connector Type: DB-15 male
+	External Reference Designator: Video
+	External Connector Type: DB-15 female
+	Port Type: Video Port
+
+Handle 0x0030, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: Keyboard
+	Internal Connector Type: PS/2
+	External Reference Designator: Keyboard
+	External Connector Type: PS/2
+	Port Type: Keyboard Port
+
+Handle 0x0031, DMI type 8, 9 bytes
+Port Connector Information
+	Internal Reference Designator: Mouse
+	Internal Connector Type: PS/2
+	External Reference Designator: Mouse
+	External Connector Type: PS/2
+	Port Type: Mouse Port
+
+Handle 0x0032, DMI type 10, 6 bytes
+On Board Device Information
+	Type: Video
+	Status: Enabled
+	Description: To Be filled by O.E.M.
+
+Handle 0x0033, DMI type 12, 5 bytes
+System Configuration Options
+	Option 1: To Be Filled By O.E.M.
+	Option 2: To Be Filled By O.E.M.
+	Option 3: To Be Filled By O.E.M.
+
+Handle 0x0034, DMI type 13, 22 bytes
+BIOS Language Information
+	Installable Languages: 1
+		enUS
+	Currently Installed Language: enUS
+
+Handle 0x0035, DMI type 18, 23 bytes
+32-bit Memory Error Information
+	Type: Bad Read
+	Granularity: Device Level
+	Operation: Read
+	Vendor Syndrome: Unknown
+	Memory Array Address: Unknown
+	Device Address: Unknown
+	Resolution: Unknown
+
+Handle 0x0036, DMI type 16, 15 bytes
+Physical Memory Array
+	Location: System Board Or Motherboard
+	Use: System Memory
+	Error Correction Type: None
+	Maximum Capacity: 3 GB
+	Error Information Handle: 0x0035
+	Number Of Devices: 4
+
+Handle 0x0037, DMI type 19, 15 bytes
+Memory Array Mapped Address
+	Starting Address: 0x00000000000
+	Ending Address: 0x100000003FF
+	Range Size: 1073741825 kB
+	Physical Array Handle: 0x0036
+	Partition Width: 0
+
+Handle 0x0038, DMI type 17, 27 bytes
+Memory Device
+	Array Handle: 0x0036
+	Error Information Handle: 0x0035
+	Total Width: 64 bits
+	Data Width: 72 bits
+	Size: 16 MB
+	Form Factor: DIMM
+	Set: None
+	Locator: DIMM1
+	Bank Locator: BANK0
+	Type: SDRAM
+	Type Detail: Synchronous
+	Speed: Unknown
+	Manufacturer: Manufacturer1
+	Serial Number: SerNum1
+	Asset Tag: AssetTagNum1
+	Part Number: PartNum1
+
+Handle 0x0039, DMI type 20, 19 bytes
+Memory Device Mapped Address
+	Starting Address: 0x00000000000
+	Ending Address: 0x000010003FF
+	Range Size: 16385 kB
+	Physical Device Handle: 0x0038
+	Memory Array Mapped Address Handle: 0x0037
+	Partition Row Position: 1
+	Interleaved Data Depth: 1
+
+Handle 0x003A, DMI type 17, 27 bytes
+Memory Device
+	Array Handle: 0x0036
+	Error Information Handle: 0x0035
+	Total Width: 64 bits
+	Data Width: 72 bits
+	Size: 16 MB
+	Form Factor: DIMM
+	Set: None
+	Locator: DIMM2
+	Bank Locator: BANK1
+	Type: SDRAM
+	Type Detail: Synchronous
+	Speed: Unknown
+	Manufacturer: Manufacturer2
+	Serial Number: SerNum2
+	Asset Tag: AssetTagNum2
+	Part Number: PartNum2
+
+Handle 0x003B, DMI type 20, 19 bytes
+Memory Device Mapped Address
+	Starting Address: 0x00001000000
+	Ending Address: 0x000020003FF
+	Range Size: 16385 kB
+	Physical Device Handle: 0x003A
+	Memory Array Mapped Address Handle: 0x0037
+	Partition Row Position: 1
+	Interleaved Data Depth: 1
+
+Handle 0x003C, DMI type 17, 27 bytes
+Memory Device
+	Array Handle: 0x0036
+	Error Information Handle: 0x0035
+	Total Width: 64 bits
+	Data Width: 72 bits
+	Size: 16 MB
+	Form Factor: DIMM
+	Set: None
+	Locator: DIMM2
+	Bank Locator: BANK1
+	Type: SDRAM
+	Type Detail: Synchronous
+	Speed: Unknown
+	Manufacturer: Manufacturer2
+	Serial Number: SerNum2
+	Asset Tag: AssetTagNum2
+	Part Number: PartNum2
+
+Handle 0x003D, DMI type 20, 19 bytes
+Memory Device Mapped Address
+	Starting Address: 0x00002000000
+	Ending Address: 0x000030003FF
+	Range Size: 16385 kB
+	Physical Device Handle: 0x003C
+	Memory Array Mapped Address Handle: 0x0037
+	Partition Row Position: 1
+	Interleaved Data Depth: 1
+
+Handle 0x003E, DMI type 17, 27 bytes
+Memory Device
+	Array Handle: 0x0036
+	Error Information Handle: 0x0035
+	Total Width: 64 bits
+	Data Width: 72 bits
+	Size: 16 MB
+	Form Factor: DIMM
+	Set: None
+	Locator: DIMM2
+	Bank Locator: BANK1
+	Type: SDRAM
+	Type Detail: Synchronous
+	Speed: Unknown
+	Manufacturer: Manufacturer2
+	Serial Number: SerNum2
+	Asset Tag: AssetTagNum2
+	Part Number: PartNum2
+
+Handle 0x003F, DMI type 20, 19 bytes
+Memory Device Mapped Address
+	Starting Address: 0x00003000000
+	Ending Address: 0x000040003FF
+	Range Size: 16385 kB
+	Physical Device Handle: 0x003E
+	Memory Array Mapped Address Handle: 0x0037
+	Partition Row Position: 1
+	Interleaved Data Depth: 1
+
+Handle 0x0040, DMI type 23, 13 bytes
+System Reset
+	Status: Disabled
+	Watchdog Timer: Not Present
+
+Handle 0x0041, DMI type 32, 20 bytes
+System Boot Information
+	Status: No errors detected
+
+Handle 0x0042, DMI type 127, 4 bytes
+End Of Table
+
+EOF
diff --git a/tests/hyperv/sbin/id b/tests/hyperv/sbin/id
new file mode 100755
index 0000000..6c704ac
--- /dev/null
+++ b/tests/hyperv/sbin/id
@@ -0,0 +1,2 @@
+#!/bin/sh -
+echo 0
diff --git a/tests/hyperv/sbin/virt-what-cpuid-helper b/tests/hyperv/sbin/virt-what-cpuid-helper
new file mode 100755
index 0000000..1604462
--- /dev/null
+++ b/tests/hyperv/sbin/virt-what-cpuid-helper
@@ -0,0 +1,3 @@
+#!/bin/sh -
+echo "Microsoft Hv"
+
diff --git a/tests/test-hyperv.sh b/tests/test-hyperv.sh
new file mode 100755
index 0000000..5015cc5
--- /dev/null
+++ b/tests/test-hyperv.sh
@@ -0,0 +1,34 @@
+# Test for Microsoft HyperV.
+# Copyright (C) 2008-2011 Red Hat Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+root=tests/hyperv
+
+output="$(./virt-what --test-root=$root 2>&1)"
+expected="hyperv"
+
+if [ "$output" != "$expected" ]; then
+    echo "$0: test failed because output did not match expected"
+    echo "Expected output was:"
+    echo "----------------------------------------"
+    echo "$expected"
+    echo "----------------------------------------"
+    echo "But the actual output of the program was:"
+    echo "----------------------------------------"
+    echo "$output"
+    echo "----------------------------------------"
+    exit 1
+fi
diff --git a/virt-what.in b/virt-what.in
index d1ac935..10798f7 100644
--- a/virt-what.in
+++ b/virt-what.in
@@ -100,8 +100,10 @@ if [ "$cpuid" = "Microsoft Hv" ]; then
 fi
 
 # Check for VirtualPC.
-
-if echo "$dmi" | grep -q 'Manufacturer: Microsoft Corporation'; then
+# The negative check for cpuid is to distinguish this from Hyper-V
+# which also has the same manufacturer string in the SM-BIOS data.
+if [ "$cpuid" != "Microsoft Hv" ] &&
+    echo "$dmi" | grep -q 'Manufacturer: Microsoft Corporation'; then
     echo virtualpc
 fi
 
diff --git a/virt-what.pod b/virt-what.pod
index 21c956a..3362bc5 100644
--- a/virt-what.pod
+++ b/virt-what.pod
@@ -29,7 +29,7 @@ don't know about or cannot detect.
 
 This is Hyper-V.
 
-Status: from MSDN description, not tested.
+Status: confirmed by RWMJ
 
 =item B<ibm_systemz>
 
-- 
1.7.1

